src/sentry/similarity/features.py

```diff
@@ -95,69 +93,96 @@ def extract(self, event):
                 )
         return results
 
-    def record(self, event):
+    def record(self, events):
+        if not events:
+            return []
+
+        scope = None
+        key = None
+
         items = []
-        for label, features in self.extract(event).items():
-            try:
-                features = map(self.encoder.dumps, features)
-            except Exception as error:
-                log = (
-                    logger.debug if isinstance(error, self.expected_encoding_errors) else
-                    functools.partial(logger.warning, exc_info=True)
-                )
-                log(
-                    'Could not encode features from %r for %r due to error: %r',
-                    event,
-                    label,
-                    error,
-                )
-            else:
-                if features:
-                    items.append((self.aliases[label], features, ))
+        for event in events:
+            for label, features in self.extract(event).items():
+                if scope is None:
+                    scope = self.__get_scope(event.project)
+                else:
+                    assert self.__get_scope(
+                        event.project
+                    ) == scope, 'all events must be associated with the same project'
+
+                if key is None:
+                    key = self.__get_key(event.group)
+                else:
+                    assert self.__get_key(
+                        event.group
+                    ) == key, 'all events must be associated with the same group'
+
+                try:
+                    features = map(self.encoder.dumps, features)
+                except Exception as error:
+                    log = (
+                        logger.debug if isinstance(error, self.expected_encoding_errors) else
+                        functools.partial(logger.warning, exc_info=True)
+                    )
+                    log(
+                        'Could not encode features from %r for %r due to error: %r',
+                        event,
+                        label,
+                        error,
+                    )
+                else:
+                    if features:
+                        items.append((self.aliases[label], features, ))
+
         return self.index.record(
-            self.__get_scope(event.project),
-            self.__get_key(event.group),
+            scope,
+            key,
             items,
```

Comment:
In the future this would be probably a good idea to make a generator instead too